community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java

```diff
@@ -272,14 +286,42 @@ public boolean shouldBeTerminated()
         return terminated;
     }
 
+    /**
+     * {@inheritDoc}
+     * <p>
+     * This method is guarded by {@link #terminationReleaseLock} to coordinate concurrent
+     * {@link #close()} and {@link #release()} calls.
+     */
     @Override
     public void markForTermination()
     {
-        if ( !terminated )
+        if ( !canBeTerminated() )
         {
-            failure = true;
-            terminated = true;
-            transactionMonitor.transactionTerminated();
+            return;
+        }
+
+        int initialReuseCount = reuseCount;
+        terminationReleaseLock.lock();
```

Comment:
Having explicit lock or lock object just makes it a bit more clear what operations is it coordinating. In this case it's name `terminationReleaseLock` tries to explain smth. I can change this to be an `Object` and use `synchronized(terminationReleaseSync)` if needed.